papers/numba/array_vs_list/array_vs_list.rst

```diff
@@ -0,0 +1,160 @@
+accesses, which limit memory bandwidth and frequently result in slower code.
+
+Numba has support for Python lists in nopython mode, with some limitations
+that improve performance.  First, all elements of a list in nopython mode must
+have the same basic numerical type: integer, floating point, or complex. Numba
+will apply its standard casting rules for math expressions to determine this
+common type.  Roughly, types of different sizes will be cast to the largest
+size, integers and floats cast up to floats, and floats and
+complex cast up to complex.  For example:
+
+.. code-block:: python
+
+    @numba.jit(nopython=True)
+    def make_list():
+        return [1, 2, 3.0]
+
+    print(make_list())
+
+Prints a list of three floats:
+
+.. code-block: python
+
+    [1.0, 2.0, 3.0]
+
+There are some drawbacks to lists in Numba, however.  Because nopython mode
+requires all arguments to be translated into machine-native forms (hence the
+"nopython" name), any Python list passed into a Numba-compiled function as an
+argument must go through an "unwrapping" step which can be quite time
+consuming for large lists.  This trivial function:
+
+.. code-block:: python
+
+    @numba.jit(nopython=True)
+    def return_first(a_list):
+        return a_list[0]
+
+is more than 100x slower in Numba than in pure Python when ``a_list`` is a
+1000 element list.
+
+As an alternative to lists, NumPy includes a typed multidimensional array
+object (which we'll call a "NumPy array" for short). NumPy arrays can be used
+in much the same way as lists in numerical code, but have several major
+benefits:
+
+  * They store their data in a native, packed format that is much more space efficient, and memory bandwidth efficient.
+  * Numba can "unwrap" a NumPy array very quickly.
+  * Multi-dimensional arrays work in Numba, but nested lists are not supported.
+
+In comparison to above, the ``return_first`` function is only 1.9x slower in
+Numba than in Python.  Numba continues to be slower in this case because the
+function body itself is so simple, the fixed overhead of function dispatch is
+dominating the runtime.
+
+Let's consider a situation where we might be tempted to use a list, and see
+how a NumPy array might be better.  A common operation when dealing with large
+amounts of data is *filtering*.  Suppose we have a million values from a
+faulty sensor, where incorrect readings appear with values less than -1000.
+Before we do any further statistics on this data, we need to remove the
+bad values.
+
+Following the advice given above, we would rightly decide to store the data in
+a NumPy array.  Passing the data in as a Python list would be tremendously
+slow due to the unwrapping overhead.  However, inside the function we don't
+know ahead of time how many values we need to emit, so we might decide to
+create an empty list and append to it:
+
+.. code-block:: python
+
+    @numba.jit(nopython=True)
+    def filter_bad_try1(values, lower, upper):
```

Comment:
Just a nit, but I feel the fact the function has "bad" in it is a bit confusing to the reader: is it bad because it has wrong behaviour, or because it is too slow? Also, even the "good" solution is stil named "bad".